### DISCUSSION TITLE
Welcome to Flipper Discussions!

### DISCUSSION THREAD

**Discussion by jnunemaker** [jnunemaker/flipper#488] [General]
Super stoked to see github has added discussions. I remember seeing them in beta years ago, then they were shelved and it looks like now they are back. 😄 

I love talking about 🐬 flipper, so please feel free to use this as a place to:

* Ask questions.
* Share ideas or example uses of flipper.
* Engage with other community members.

To get started, **comment below with an introduction of yourself** and tell everyone how you use flipper (if you can).

## My Flipper Story

I'll start. I'm John. 

### rollout 

I started with feature flags/toggles in 2008-ish through https://github.com/FetLife/rollout while working on words with friends. Every time I tried to roll out this new caching mechanism, the site would fall over (cold cache problem). 

We plopped in rollout and did a % of actors based rollout and it went great. Everything after that went behind a feature flag. Eventually we were such heavy users of features that our redis instance wasn't a fan of us. We forked rollout and put memcached in front of the redis bits. There were a few other niggling issues with rollout, but it worked great.

### flipper 

One weekend in 2013, I wrote my own version based on adapters and flipper was born. Need to wrap redis with memcache and wrap that in memory? No problem, just whip some adapters together (or now just piece together what is already supported out of the box). I wrote it for fun, but felt it was solid. I released it and people started using it, but I actually wasn't. 

### github.com 

I was working on this here site you are reading this on (github) and pretty much all of our releases involved deploys. Then, around 2015, flipper was rolled into github.com (might still be as of this writing) by @rsanheim and @aroben while I was on paternity leave. 

When I last checked, while still working at github, flipper was doing billions of feature checks a day. It had zero impact to overall performance and really opened up the way we released code there. It also really freed up the deploy queue for a few years as releases were done with flipper changes instead of full deploys.

### most memorable use 

One of my favorite examples of flipper was at one of the github universe events. @defunkt was CEO at the time and immediately after he announced each new feature, they were live. Seeing the audit logs of the features being enabled in chat as the presentation went on was really neat. I remember a few tweets from people amazed at how instantaneously the features were being released with the words he was saying.

### the next big thing

Around 2018 I started thinking that a hosted version of flipper would be cool based on what I saw at GitHub. I started working on it and got it functional, but then kind of sidelined it. We used it at [boxoutsports.com](https://boxoutsports.com) and on [speakerdeck.com](https://speakerdeck.com), but didn't really let anyone else in. 

Finally, we've decided it is time to polish up the last few bits and release it. We are really close. You can checkout [flippercloud.io](https://www.flippercloud.io) if you are curious.

### the end 

So that is my flipper story. I've used feature toggles to roll out new caches, to warm up and completely swap memcached clusters, to turn on logging/debugging while working on multi-datacenter stuff, to release complete redesigns of apps and to wrap up every new feature in any app I work on. 

I'm a big fan. And I'd love to hear from all of you. 

So tell me:

* How did you get started with flipper? 
* How do you use it now? 
* What do you love or hate about it? 

... or anything else that is on your mind about releasing software. 🍻 Cheers!

**Comment by gertig** [comment]
@jnunemaker any thoughts on best way to manage a set of features for an Actor in a client side SPA? Is directly hitting the API the best approach?

**Reply by jnunemaker** [reply]
I don't currently have any single page apps, so its hard for me to say. 

What we typically do when we need features in JS is to send whether its enabled or not to js along with whatever we are doing. 

Others send a big hash of all features (or all public features) and if they are enabled or not. 

Any approach that makes them available seems good. I've also seen a few posts on flipper + react. Might be out of date now but this was product hunt's approach at one point. https://blog.rstankov.com/feature-flags-in-react/

I'd love to know what others are doing and to collect some best practices.

**Reply by gertig** [reply]
Nice.

It would be great to render the features that are enabled for an actor alongside it in an API response in order to bootstrap a client that does not maintain a cookie based session. Mobile apps for example could benefit from this.

Is there perhaps a DSL for this that I missed? https://github.com/jnunemaker/flipper/blob/master/docs/api/README.md#check-if-features-are-enabled-for-an-actor

**Reply by jnunemaker** [reply]
Was on a long vacation. :) Thus the delay in response.

The HTTP API is more of an API for accessing the stored data than determining if something is enabled. We could make an endpoint for people to hit with an actor id. That would work for all the gates but the groups gate. 

What most people do in SPA's is create a Hash of feature key => true/false. You can do this with the Flipper instance by mapping all the features. If you are using preloading + memoization, that ensures that only one network call or so will be made to check the actor for all features. 

If you need an example, let me know and I can create a new one in the examples directory. 

Does any of the above help?
